GitHub Actions unable to connect to EBRAINS KG when running tests

Hello,
We, at TVB, integrated siibra in our source code to be able to extract structural and functional connectivities from the KG. We also added some pytests for this new feature.
When running the tests locally, with the EBRAINS token set as an environment variable, everything works fine. But when the tests are run in GitHub Actions, the calls to the KG return a 401 status code:
==================================== ERRORS ====================================
```
__ ERROR at setup of TestSiibraBase.test_filter_structural_connectivity_by_id __

self = <tvb.tests.framework.adapters.creators.siibra_base_test.TestSiibraBase object at 0x7f56fe772980>
create_test_atlases_and_parcellations = None

    @pytest.fixture()
    def create_weights_and_tracts(self, create_test_atlases_and_parcellations):
        """
        Return all the weights and tracts available in siibra for default atlas and parcellation
        """
>       weights = siibra.get_features(self.julich_parcellation, siibra.modalities.StreamlineCounts)

tvb/tests/framework/adapters/creators/siibra_base_test.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../.local/lib/python3.10/site-packages/siibra/features/query.py:94: in get_features
    cls._instances[querytype, args_hash] = querytype(**kwargs)
../../../../.local/lib/python3.10/site-packages/siibra/features/connectivity.py:333: in __init__
    HcpConnectivityFetcher.__init__(self, filename_keyword="Counts")
../../../../.local/lib/python3.10/site-packages/siibra/features/connectivity.py:205: in __init__
    self._connector = EbrainsPublicDatasetConnector(
../../../../.local/lib/python3.10/site-packages/siibra/retrieval/repositories.py:286: in __init__
    result = EbrainsRequest(url, DECODERS[".json"]).get()
../../../../.local/lib/python3.10/site-packages/siibra/retrieval/requests.py:326: in get
    return super().get()
../../../../.local/lib/python3.10/site-packages/siibra/retrieval/requests.py:163: in get
    data = self._retrieve()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <siibra.retrieval.requests.EbrainsRequest object at 0x7f56ea7edc90>

    def _retrieve(self):
        # Loads the data from http if required.
        # If the data is already cached, None is returned,
        # otherwise data (as it is already in memory anyway).
        # The caller should load the cachefile only
        # if None is returned.
    
        if self.cached and not self.refresh:
            # in cache. Just load the file
            logger.debug(
                f"Already in cache at {os.path.basename(self.cachefile)}: {self.url}"
            )
            return
        else:
            # not yet in cache, perform http request.
            logger.debug(f"Loading {self.url} to {os.path.basename(self.cachefile)}")
            if self.msg_if_not_cached is not None:
                logger.info(self.msg_if_not_cached)
            headers = self.kwargs.get('headers', {})
            other_kwargs = {key: self.kwargs[key] for key in self.kwargs if key != "headers"}
            if self.post:
                r = requests.post(self.url, headers={
                    **USER_AGENT_HEADER,
                    **headers,
                }, **other_kwargs)
            else:
                r = requests.get(self.url, headers={
                    **USER_AGENT_HEADER,
                    **headers,
                }, **other_kwargs)
            if r.ok:
                with open(self.cachefile, "wb") as f:
                    f.write(r.content)
                self.refresh = False
                return r.content
            else:
>               raise SiibraHttpRequestError(r)
E               siibra.retrieval.requests.SiibraHttpRequestError: Cannot execute http request.
E                   Status code: 401
E                   Url:         https://core.kg.ebrains.eu/v3-beta/queries//bebbe365-a0d6-41ea-9ff8-2554c15f70b7/instances?stage=RELEASED&dataset_id=0f1ccc4a-9a11-4697-b43f-9c9c8ac543e6
```
Please refer to this [GitHub action run](https://github.com/the-virtual-brain/tvb-root/runs/7803372307?check_suite_focus=true) and this [PR](https://github.com/the-virtual-brain/tvb-root/pull/585) for more information.
_Note:_ In the actions run page you will see printed out a test ebrains token, but even with a valid token the problem persists.

Do you have any suggestions on how to overcome this issue to be able to successfully run the tests from GitHub Actions?

--- a/.github/workflows/siibra-testing.yml
+++ b/.github/workflows/siibra-testing.yml
@@ -47,8 +47,9 @@ jobs:
   e2e-tests:
     runs-on: ${{ matrix.os }}
     env:
-      KEYCLOAK_CLIENT_ID: ${{ secrets.KEYCLOAK_CLIENT_ID }}
-      KEYCLOAK_CLIENT_SECRET: ${{ secrets.KEYCLOAK_CLIENT_SECRET }}
+      # KEYCLOAK_CLIENT_ID: ${{ secrets.KEYCLOAK_CLIENT_ID }}
+      # KEYCLOAK_CLIENT_SECRET: ${{ secrets.KEYCLOAK_CLIENT_SECRET }}
+      FOO: BAR
     strategy:
       fail-fast: false
       matrix:

--- a/e2e/features/external/test_ebrains.py
+++ b/e2e/features/external/test_ebrains.py
@@ -2,7 +2,7 @@ import siibra
 import pytest
 
 concepts = [
-    (siibra.get_region("julich 2.9", "v2"), siibra.features.external.EbrainsDataset),
+    (siibra.get_region("julich 2.9", "v2"), siibra.features.external.EbrainsDataFeature),
     (siibra.get_region("julich 2.9", "v2"), "ebrains"),
 ]
 

--- a/examples/03_data_features/000_matchings.py
+++ b/examples/03_data_features/000_matchings.py
@@ -39,7 +39,7 @@ print(siibra.features.Feature.SUBCLASSES) # TODO: This is a temporary fix.
 # As a first example, we use brain region V2 to query for "EbrainsRegionalDataset" features.
 # See :ref:`ebrains_datasets` for more information about this modality.
 v2 = siibra.get_region("julich 2.9", "v2")
-features = siibra.features.get(v2, siibra.features.external.EbrainsDataset)
+features = siibra.features.get(v2, siibra.features.external.EbrainsDataFeature)
 for feature in features:
     print(f" - {feature.name}")
 

--- a/examples/03_data_features/005_ebrains_datasets.py
+++ b/examples/03_data_features/005_ebrains_datasets.py
@@ -30,7 +30,7 @@ The returned features can thus actually have different modalities, and might inc
 import siibra
 atlas = siibra.atlases.MULTILEVEL_HUMAN_ATLAS
 region = atlas.get_region("v2")
-features = siibra.features.get(region, siibra.features.external.EbrainsDataset)
+features = siibra.features.get(region, siibra.features.external.EbrainsDataFeature)
 for feature in features:
    print(f" - {feature.name}")
 

--- a/siibra/features/external/__init__.py
+++ b/siibra/features/external/__init__.py
@@ -13,9 +13,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from .ebrains import EbrainsDataset
+from .ebrains import EbrainsDataFeature
 
 def __dir__():
     return [
-        "EbrainsDataset",
+        "EbrainsDataFeature",
     ]

--- a/siibra/features/external/ebrains.py
+++ b/siibra/features/external/ebrains.py
@@ -21,7 +21,7 @@ from ..basetypes import feature
 from ...retrieval import datasets
 
 
-class EbrainsDataset(feature.Feature, datasets.EbrainsDataset):
+class EbrainsDataFeature(feature.Feature, datasets.EbrainsDataset):
 
     def __init__(
         self,

--- a/siibra/livequeries/ebrains.py
+++ b/siibra/livequeries/ebrains.py
@@ -28,7 +28,7 @@ from tqdm import tqdm
 from tempfile import NamedTemporaryFile
 
 
-class EbrainsFeatureQuery(query.LiveQuery, args=[], FeatureType=_ebrains.EbrainsDataset):
+class EbrainsFeatureQuery(query.LiveQuery, args=[], FeatureType=_ebrains.EbrainsDataFeature):
 
     # in EBRAINS knowledge graph prior to v3, versions were modelled
     # in dataset names. Typically found formats are (v1.0) and [rat, v2.1]
@@ -93,7 +93,7 @@ class EbrainsFeatureQuery(query.LiveQuery, args=[], FeatureType=_ebrains.Ebrains
                     if any(e.lower() in ds_name.lower() for e in self._BLACKLIST):
                         continue
 
-                dset = _ebrains.EbrainsDataset(
+                dset = _ebrains.EbrainsDataFeature(
                     dataset_id=ds_id,
                     name=ds_name,
                     anchor=_anchor.AnatomicalAnchor(

--- a/siibra/livequeries/ebrains.py
+++ b/siibra/livequeries/ebrains.py
@@ -43,10 +43,17 @@ class EbrainsFeatureQuery(query.LiveQuery, args=[], FeatureType=_ebrains.Ebrains
         "Automated Anatomical Labeling (AAL1) atlas",
     }
 
-    loader = requests.EbrainsKgQuery(
-        query_id="siibra-kg-feature-summary-0_0_4",
-        schema="parcellationregion",
-        params={"vocab": "https://schema.hbp.eu/myQuery/"},
+    loader = requests.MultiSourcedRequest(
+        requests=[
+            requests.GitlabProxy(
+                flavour=requests.GitlabProxyEnum.PARCELLATIONREGION_V1,
+            ),
+            requests.EbrainsKgQuery(
+                query_id="siibra-kg-feature-summary-0_0_4",
+                schema="parcellationregion",
+                params={"vocab": "https://schema.hbp.eu/myQuery/"},
+            )
+        ]
     )
 
     parcellation_ids = None

--- a/siibra/retrieval/datasets.py
+++ b/siibra/retrieval/datasets.py
@@ -14,7 +14,7 @@
 # limitations under the License.
 
 
-from .requests import EbrainsKgQuery
+from .requests import EbrainsKgQuery, MultiSourcedRequest, GitlabProxy, GitlabProxyEnum
 
 import re
 from typing import Union, List
@@ -69,10 +69,19 @@ class EbrainsDataset:
     def detail(self):
         if not self._cached_data:
             match = re.search(r"([a-f0-9-]+)$", self.id)
-            self._cached_data = EbrainsKgQuery(
-                query_id="interactiveViewerKgQuery-v1_0",
-                instance_id=match.group(1),
-                params={"vocab": "https://schema.hbp.eu/myQuery/"},
+            instance_id = match.group(1)
+            self._cached_data = MultiSourcedRequest(
+                requests=[
+                    GitlabProxy(
+                        GitlabProxyEnum.DATASET_V1,
+                        instance_id=instance_id,
+                    ),
+                    EbrainsKgQuery(
+                        query_id="interactiveViewerKgQuery-v1_0",
+                        instance_id=instance_id,
+                        params={"vocab": "https://schema.hbp.eu/myQuery/"},
+                    )
+                ]
             ).data
         return self._cached_data
 

--- a/siibra/retrieval/repositories.py
+++ b/siibra/retrieval/repositories.py
@@ -14,6 +14,7 @@
 # limitations under the License.
 
 from .requests import DECODERS, HttpRequest, EbrainsRequest, SiibraHttpRequestError
+from .cache import CACHE
 
 from .. import logger
 
@@ -160,7 +161,12 @@ class LocalFileRepository(RepositoryConnector):
 
 class GitlabConnector(RepositoryConnector):
 
-    def __init__(self, server: str, project: int, reftag: str, skip_branchtest=False):
+    def __init__(self, server: str, project: int, reftag: str, skip_branchtest=False, *, archive_mode=False):
+        """
+        archive_mode: in archive mode, the entire repository is downloaded as an archive. This is necessary/could be useful for repositories with numerous files.
+            n.b. only archive_mode should only be set for trusted domains. Extraction of archive can result in files created outside the path
+            see https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractall
+        """
         # TODO: the query builder needs to check wether the reftag is a branch, and then not cache.
         assert server.startswith("http")
         RepositoryConnector.__init__(
@@ -173,6 +179,7 @@ class GitlabConnector(RepositoryConnector):
         )
         self._tag_checked = True if skip_branchtest else False
         self._want_commit_cached = None
+        self.archive_mode = archive_mode
 
     def __str__(self):
         return f"{self.__class__.__name__} {self.base_url} {self.reftag}"
@@ -230,6 +237,38 @@ class GitlabConnector(RepositoryConnector):
             for e in results
             if e["type"] == "blob" and e["name"].endswith(end)
         ]
+    
+    def get(self, filename, folder="", decode_func=None):
+        if not self.archive_mode:
+            return super().get(filename, folder, decode_func)
+        
+        ref = self.reftag if self.want_commit is None else self.want_commit["short_id"]
+        archive_directory = CACHE.build_filename(self.base_url + ref) if self.archive_mode else None
+        import os, requests
+        
+        if not os.path.isdir(archive_directory):
+            
+            url = self.base_url + f"/archive.tar.gz?sha={ref}"
+            resp = requests.get(url)
+            tar_filename = f"{archive_directory}.tar.gz"
+
+            resp.raise_for_status()
+            with open(tar_filename, "wb") as fp:
+                fp.write(resp.content)
+            
+            import tarfile
+            tar = tarfile.open(tar_filename, "r:gz")
+            tar.extractall(archive_directory)
+            for _dir in os.listdir(archive_directory):
+                for file in os.listdir(f"{archive_directory}/{_dir}"):
+                    os.rename(f"{archive_directory}/{_dir}/{file}", f"{archive_directory}/{file}")
+                os.rmdir(f"{archive_directory}/{_dir}")
+            
+        suitable_decoders = [dec for sfx, dec in DECODERS.items() if filename.endswith(sfx)]
+        decoder = suitable_decoders[0] if len(suitable_decoders) > 0 else lambda b: b
+
+        with open(f"{archive_directory}/{folder}/{filename}", "rb") as fp:
+            return decoder(fp.read())
 
 
 class ZipfileConnector(RepositoryConnector):

--- a/siibra/retrieval/requests.py
+++ b/siibra/retrieval/requests.py
@@ -32,7 +32,7 @@ import urllib
 import pandas as pd
 import numpy as np
 from tqdm import tqdm
-from typing import List, TYPE_CHECKING
+from typing import List, Callable, Any, TYPE_CHECKING
 from enum import Enum
 from functools import wraps
 
@@ -514,7 +514,7 @@ class GitlabProxy(HttpRequest):
         GitlabProxyEnum.PARCELLATIONREGION_V1: "ebrainsquery/v1/parcellationregions",
     }
 
-    def __init__(self, flavour: GitlabProxyEnum, instance_id=None, postprocess=lambda results: {"results": results}):
+    def __init__(self, flavour: GitlabProxyEnum, instance_id=None, postprocess: Callable[['GitlabProxy', Any], Any]=(lambda proxy, obj: obj if hasattr(proxy, "instance_id") and proxy.instance_id else { "results": obj })):
         if flavour not in GitlabProxyEnum:
             raise RuntimeError(f"Can only proxy enum members")
 
@@ -527,8 +527,8 @@ class GitlabProxy(HttpRequest):
 
     def get(self):
         if self.instance_id:
-            return self.postprocess(self.flavour.get(f"{self.folder}/{self.instance_id}.json"))
-        return self.postprocess(self.flavour.get(f"{self.folder}/_all.json"))
+            return self.postprocess(self, self.flavour.get(f"{self.folder}/{self.instance_id}.json"))
+        return self.postprocess(self, self.flavour.get(f"{self.folder}/_all.json"))
     
 
 class MultiSourceRequestException(Exception): pass

--- a/siibra/retrieval/requests.py
+++ b/siibra/retrieval/requests.py
@@ -32,6 +32,12 @@ import urllib
 import pandas as pd
 import numpy as np
 from tqdm import tqdm
+from typing import List, TYPE_CHECKING
+from enum import Enum
+from functools import wraps
+
+if TYPE_CHECKING:
+    from .repositories import GitlabConnector
 
 USER_AGENT_HEADER = {"User-Agent": f"siibra-python/{__version__}"}
 
@@ -458,3 +464,91 @@ class EbrainsKgQuery(EbrainsRequest):
                     f"URL was: {e.url}"
                 )
         return result
+
+def try_all_connectors():
+    def outer(fn):
+        @wraps(fn)
+        def inner(self: 'GitlabProxyEnum', *args, **kwargs):
+            exceptions = []
+            for connector in self.connectors:
+                try:
+                    return fn(self, *args, connector=connector, **kwargs)
+                except Exception as e:
+                    exceptions.append(e)
+            else:
+                for exc in exceptions:
+                    logger.error(exc)
+                raise Exception("try_all_connectors failed")
+        return inner
+    return outer
+
+class GitlabProxyEnum(Enum):
+    DATASET_V1="DATASET_V1"
+    PARCELLATIONREGION_V1="PARCELLATIONREGION_V1"
+    DATASET_V3="DATASET_V3"
+
+    @property
+    def connectors(self) -> List['GitlabConnector']:
+        servers = [
+            ("https://jugit.fz-juelich.de", 7846),
+            ("https://gitlab.ebrains.eu", 421),
+        ]
+        from .repositories import GitlabConnector
+        return [GitlabConnector(server[0], server[1], "master", archive_mode=True) for server in servers]
+
+    @try_all_connectors()
+    def search_files(self, folder: str, suffix=None, recursive=True, *, connector: 'GitlabConnector'=None) -> List[str]:
+        assert connector
+        return connector.search_files(folder, suffix=suffix, recursive=recursive)
+
+    @try_all_connectors()
+    def get(self, filename, decode_func=None, *, connector: 'GitlabConnector'=None):
+        assert connector
+        return connector.get(filename, "", decode_func)
+
+class GitlabProxy(HttpRequest):
+
+    folder_dict = {
+        GitlabProxyEnum.DATASET_V1: "ebrainsquery/v1/datasets",
+        GitlabProxyEnum.DATASET_V3: "ebrainsquery/v3/datasets",
+        GitlabProxyEnum.PARCELLATIONREGION_V1: "ebrainsquery/v1/parcellationregions",
+    }
+
+    def __init__(self, flavour: GitlabProxyEnum, instance_id=None, postprocess=lambda results: {"results": results}):
+        if flavour not in GitlabProxyEnum:
+            raise RuntimeError(f"Can only proxy enum members")
+
+        self.flavour = flavour
+        self.folder = self.folder_dict[flavour]
+        self.postprocess = postprocess
+        self.instance_id = instance_id
+        self._cached_files = None
+
+
+    def get(self):
+        if self.instance_id:
+            return self.postprocess(self.flavour.get(f"{self.folder}/{self.instance_id}.json"))
+        return self.postprocess(self.flavour.get(f"{self.folder}/_all.json"))
+    
+
+class MultiSourceRequestException(Exception): pass
+
+class MultiSourcedRequest:
+    requests: List[HttpRequest] = []
+    
+    def __init__(self, requests: List[HttpRequest]) -> None:
+        self.requests = requests
+    
+    def get(self):
+        exceptions = []
+        for req in self.requests:
+            try:
+                return req.get()
+            except Exception as e:
+                exceptions.append(e)
+        else:
+            raise MultiSourceRequestException("All requests failed:\n" + "\n".join(str(exc) for exc in exceptions))
+
+    @property
+    def data(self):
+        return self.get()

--- a/test/features/external/test_ebrains.py
+++ b/test/features/external/test_ebrains.py
@@ -1,11 +1,11 @@
 import unittest
 from unittest.mock import PropertyMock, patch
-from siibra.features.external.ebrains import EbrainsDataset, feature, datasets
+from siibra.features.external.ebrains import EbrainsDataFeature, feature
 
 class TestEbrains(unittest.TestCase):
     @staticmethod
     def get_instance(dataset_id=None):
-        return EbrainsDataset(
+        return EbrainsDataFeature(
             dataset_id=dataset_id,
             anchor=None,
             name="foo bar",

--- a/test/features/test_ebrainsquery.py
+++ b/test/features/test_ebrainsquery.py
@@ -11,7 +11,7 @@ class TestEbrainsQuery(unittest.TestCase):
     def setUpClass(cls):
         atlas = siibra.atlases.MULTILEVEL_HUMAN_ATLAS
         region = atlas.get_region("hoc1 left")
-        cls.feat = siibra.features.get(region, siibra.features.external.EbrainsDataset)
+        cls.feat = siibra.features.get(region, siibra.features.external.EbrainsDataFeature)
 
     def test_some_result_returned(self):
         assert len(self.feat) > 0

